### PR TITLE
bpo-35431: Add a function computing binomial numbers to the math module

### PR DESCRIPTION
**Pull request by KellerFuchs**
- Add tests specifying the behaviour of `math.binomial(k, n)`:
  - [x] when `0 <= k <= n`;
  - [x] when `k > n`;
  - [x] when `k` or `n` are negative or non-integral;
  - [x] when `k` or `n` are neither `int` nor `float`.
- [x] Implement `math.binomial` in C; done by @FR4NKESTI3N 
- [x] ~Profile/benchmark and optimize if needed.~
      @rhettinger suggested optimizing this, if needed, in a later PR.

https://bugs.python.org/issue35431

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by KellerFuchs** [comment]
Note that this is a work in progress; currently, it only contains tests specifying the behaviour of the function, so we can more easily discuss what's the desired behaviour.

File: Lib/test/test_math.py [hunk 1]
```diff
@@ -523,6 +523,52 @@ def testFactorialHugeInputs(self):
...
+    def testBinomialOne(self):
+        """Test (n choose 0) = (n choose n) = 1"""
+        for n in range(100):
+            self.assertEqual(1, binomial(n, 0))
+            self.assertEqual(1, binomial(n, n))
+
+    def testBinomialValueErrors(self):
+        """Test that math.binomial raises ValueError on negative inputs."""
+        self.assertRaises(ValueError, math.binomial, 0, -1)
+        self.assertRaises(ValueError, math.binomial, 0, -1.0)
```
Comment (mdickinson) [inline Lib/test/test_math.py]: There are two possible causes of exception here: the negative value, and the use of `float` type (assuming that there's a decision not to accept `float`s, at least for now). I think the `TypeError` for the `float` should win in this case.

**Comment by rhettinger** [comment]
Let's name this *comb()* instead of *binomial()* please.

**Comment by FR4NKESTI3N** [comment]
Can I work on C implementation if noone else is working on it right now?

File: Lib/test/test_math.py (hunk 1 above)
Comment (KellerFuchs) [reply Lib/test/test_math.py]: Yeah, I hadn't yet updated the tests following the discussion.

**Comment by FR4NKESTI3N** [comment]
Can we work on PR #11414 since it already passes all tests? Added you as a collaborator.

File: Modules/mathmodule.c [hunk 2]
```diff
@@ -2490,6 +2490,151 @@ math_isclose_impl(PyObject *module, double a, double b, double rel_tol,
...
+        dump_var = NULL;
+    }
+
+    terms = PyLong_AsLongLongAndOverflow(k, &overflow);
+    if (terms == -1 && PyErr_Occurred()) {
+        goto fail_comb;
+    }
+    else if (overflow == 1) {
+        PyErr_Format(PyExc_OverflowError,
+                     "(n - k) and k must not exceed %lld",
```
Comment (FR4NKESTI3N) [inline Modules/mathmodule.c]: It has to be minimum of the two since the algorithm uses symmetry to reduce no of iterations

**Comment by FR4NKESTI3N** [comment]
Also, Is it okay to have multiple methods for tests of a single function? All other function in math have their tests encased in a single method which seems easier to maintain.

File: Modules/mathmodule.c (hunk 2 above)
Comment (KellerFuchs) [reply Modules/mathmodule.c]: Ah, right.

**Comment by KellerFuchs** [comment]
> Is it okay to have multiple methods for tests of a single function?

I would assume as much: I modelled those tests after the ones for `factorial`, which follow exactly that model (single function per property, with an informative name)

**Comment by KellerFuchs** [comment]
@FR4NKESTI3N The main reasons tests are failing is that the I had to reconcile the testsuite in one branch with the implemented behavior in the other, which would have to be done regardless of which branch we are pushing with.
It's now done (I had missed one test where `k>n`, and tests now pass.

I pushed to this PR rather than #11414 because this was the one I had write access to  :)

**Comment by KellerFuchs** [comment]
@csabella, @pablogsal, @scoder: Taking the liberty to ping you, since you commented on #11414 

This is the PR and testsuite I started during the discussion on `bugs.python.org`, and I merged in @FR4NKESTI3N's C implementation that you already reviewed.  The principal differences are:
- minor improvements to the exception messages;
- the testsuite for `math.comb` covers some extra edge cases, and is split in independent functions (as for `math.factorial`), which are documented by their docstring.

**Comment by FR4NKESTI3N** [comment]
#11414  was already conflict free and had function name as comb, you needlessly worked by naming and renaming etc. The changes that you made in last 7-8 commits were already there. You were already invited as collaborator. 
The test method names still need to be changed to testComb. The factorial test suite has 3 functions and is pretty old, and one of those methods needed a decorator.  6+ methods are bound to cause maintenance problems in the future.
Some exception tests are still missing and "combinations" isstill written everywhere except tests. Renaming and writing them all will be useless effort as all that is already done. 
I would still insist you to work on that PR or merge that branch here again. Or maybe make the changes in that branch and then merge it with this. There were just some changes in 2 or 3 comments so it will be much more simpler to work on that branch.

**Comment by KellerFuchs** [comment]
@FR4NKESTI3N 
> #11414 was already conflict free and had function name as comb, you needlessly worked by naming and renaming etc. The changes that you made in last 7-8 commits were already there. You were already invited as collaborator.

Regarding “the last 7-8 commits” you talk of, I have no idea what you are going off, given that all I added since the initial merge 4 days ago (before you commented on this PR or added me as a collaborator) were commits afb3d36 (trivial changes, fixes the testsuite) and e4942bf (trivial rename).

Not sure what you mean by “conflict-free” since there's no conflict between this and `master` either (nor are there ever been).

On the other hand, recreating all that work on your PR would indeed be unnecessary work.

> The test method names still need to be changed to testComb

Not sure that's entirely necessary, but I just did that to avoid having an argument, since it's a quick search-and-replace.

> The factorial test suite has 3 functions and is pretty old, and one of those methods needed a decorator. 6+ methods are bound to cause maintenance problems in the future.

I am absolutely counfunded why you would expect a single function, testing a bunch of unrelated properties, to be more maintainable than separate tests.

I assure you, cramming everything into a single test does not make for maintainable testsuites.

> Some exception tests are still missing

Which ones?  I am pretty sure I kept them all when merging both testsuites.

> "combinations" isstill written everywhere except tests. Renaming and writing them all will be useless effort as all that is already done.
Clearly not:
```
$ grep -i combinations Modules/mathmodule.c Lib/test/test_math.py
[Empty output, no match found]
```

> I would still insist you to work on that PR or merge that branch here again. Or maybe make the changes in that branch and then merge it with this.

All the things you mentionned existed all along (with the exception of a single search-and-replace for test names), so I'm not sure why you want me to recreate all that on the other PR.

> There were just some changes in 2 or 3 comments so it will be much more simpler to work on that branch.

There were changes to the testsuite, which was the whole point of that merge in the first place.

**Comment by FR4NKESTI3N** [comment]
Hmm yeah, I just wanted to say said that it was wasteful. I think I was looking at an earlier commit, so I though that re-merging would be better.
Anyway fixed the name in other files.
Still not sure about unittests though but I am no expert.
Should we start on optimizations?

**Comment by KellerFuchs** [comment]
@FR4NKESTI3N I think `math.perm` ought to be a separate PR.
Regarding optimisations, like @rhettinger  I'd rather get this reviewed/merged before we submit optimized code, but it doesn't mean we can't already do the legwork while this goes through the review.

Also thanks for noticing I had missed the documentation page  :)

**Comment by KellerFuchs** [comment]
PS: Just pushed a commit documenting under which circumstances `math.comb` may raise `OverflowError`.

**Comment by FR4NKESTI3N** [comment]
Having LLONG_MAX doesn't seem right in documentation, maybe we can use something along the lines of "size of long long provided by compiler". This detail is purely implementation based anyway, so I don't know if it should even be there in documentation.

**Comment by KellerFuchs** [comment]
@FR4NKESTI3N Good point; we might just mention there is a size beyond which `math.comb` is guaranteed to raise `OverflowError`.

The documentation for the `math` module mentions that things may raise `OverflowError`, but that's in the context of `float` overflow, which is a quite different beast altogether.

File: Doc/library/math.rst [hunk 3]
```diff
@@ -206,6 +206,22 @@ Number-theoretic and representation functions
...
+   Return the binomial coefficient indexed by the pair of integers n >= k >= 0.
+
+   It is the coefficient of kth term in polynomial expansion of the expression
+   (1 + x)^n. It is also known as the number of ways to choose an unordered
+   subset of k elements from a fixed set of n elements, usually called
+   *n choose k*.
+
+   Raises :exc:`TypeError` if argument(s) are non-integer, :exc:`ValueError`
+   if argument(s) are negative or k > n, and :exc:`OverflowError` if k and (n-k)
+   are too large (beyond `LLONG_MAX`).
```
Comment (KellerFuchs) [inline Doc/library/math.rst:232]:
```suggestion
   are too large; the max. value is implementation- and platform-dependent.
```

Comment (KellerFuchs) [reply Doc/library/math.rst:232]: @FR4NKESTI3N Does that seem like reasonable wording?

**Comment by FR4NKESTI3N** [comment]
@KellerFuchs R.Hettinger reviewed the other PR so I have made the changes there. There was some change in implementation so I haven't written anything in docs about OverflowError for until Hettinger decides on implementation.

**Comment by brettcannon** [comment]
Please note that the CLA has still not been signed. Without it we cannot accept this PR.

**Comment by KellerFuchs** [comment]
@brettcannon Oh, thanks for the reminder, looks like the CLA signing thing didn't go through properly the first time around (I had some issues with DocuSign); I just redid the thing, so it should be good tomorrow.

**Comment by KellerFuchs** [comment]
PS: Also rebased the branch to deal with the merge conflist.

File: Doc/library/math.rst [hunk 4]
```diff
@@ -218,6 +218,22 @@ Number-theoretic and representation functions
    :meth:`x.__trunc__() <object.__trunc__>`.
 
 
+.. function:: comb(n, k)
+
+   Return the binomial coefficient indexed by the pair of integers n >= k >= 0.
+
+   It is the coefficient of kth term in polynomial expansion of the expression
+   (1 + x)^n. It is also known as the number of ways to choose an unordered
```
Comment (vadmium) [inline Doc/library/math.rst:226]: Other parts of this file use double asterisks (**) for exponentiation, which is less likely to be confused with other meanings such as the XOR operator.

**Comment by rhettinger** [comment]
Closing in favor of PR 11414.  Any remaining ideas from this PR can be brought up in a separate PR.

Thanks for the parallel efforts.

**Pull request closed without merging**